izpack-compiler/src/main/java/com/izforge/izpack/compiler/CompilerConfig.java

```diff
@@ -2751,6 +2751,12 @@ else if ("always".equalsIgnoreCase(content))
         {
             logger.info("Adding uninstaller");
 
+            // Write the manifest first so that it is not overwritten
+            try {
+                writeManifest();
+            } catch (IOException e) {
+                throw new CompilerException("Error while writing manifest", e);
+            }
             //REFACTOR Change the way uninstaller is created
             mergeManager.addResourceToMerge("com/izforge/izpack/uninstaller/");
             mergeManager.addResourceToMerge("uninstaller-META-INF/");
```

Comment:
@hiteshbosamiya What happens if the `/uninstaller-META-INF/MANIFEST.MF` exists and the `MANIFEST.MF` is enhanced with additional entries in `writeManifest()` and then in line 2762 the `uninstaller-META-INF/`is added to mergeManager. Will the existing `MANIFEST.MF` override the enhanced `MANIFEST.MF` or will the enhanced file stay in place?

Reply:
It takes `/uninstaller-META-INF/MANIFEST.MF` file and then appends the entries if they don't exist in `/uninstaller-META-INF/MANIFEST.MF`. This is to make sure that whatever IzPack has decided as the values don't get overwritten by the installer creator. The example is, Main-Class, we don't want anyone to overwrite it.

Reply:
That was clear to me :-)

It's the line 2762 `mergeManager.addResourceToMerge("uninstaller-META-INF/");` that adds the contents of the directory `uninstaller-META-INF` to the `mergeableList` of `MergeManagerImpl` and I wonder if the `MANIFEST.MF` is added in the twice in the list if the second overrides the changes of the `MANIFEST.MF` added in the `writeManifest()` step in line 2756.

I cannot run tests currently myself, but if you create a directory and manifest file `/uninstaller-META-INF/MANIFEST.MF` in the resources directory so that the test will pick up it in `CompilerConfig` line 2762, will the test still pass?

Reply:
Yes, currently that is how the test case is. It has `/uninstaller-META-INF/MANIFEST.MF` file (with `Manifest-Version`, `Created-By`, and `Main-Class` entries) and new method `writeManifest()` file picks up that file, adds the extra entries and then adds as a resource to merge. First, I had used `writeManifest()` method, after line no. 2762 but it didn't append the additional manifest entries. Which means once the file is written it does not get overwritten. This is also mentioned in a comment in `PakckagerBase.java` at line no. 420. I have given the similar comment at line no. 2754 of `CompilerConfig.java`.

Reply:
OK, thanks for this info. Then I'm fine with it. Thank you.